[Bug]:  "Content block is not a text block" intermittently on v1.88.0-rc.1 with Bedrock Converse + CRIS + Thinking (multi-turn)

### Check for existing issues

- [x] I have searched the existing issues and checked that my issue is not a duplicate.

### What happened?


## What happened?

Claude Code CLI (v2.1.159) intermittently throws `Error: Content block is not a text block` on the **2nd turn** of a multi-turn conversation when using Bedrock Converse API with Cross-Region Inference (CRIS) and extended thinking enabled.

The first turn always works. The error occurs when the conversation history already contains thinking blocks from the previous assistant response.

## Relevant Issues

- Related to #24641 (document content blocks silently dropped during message conversion)
- Related to #20370 (empty text content blocks not sanitized)
- Claude Code issue: anthropics/claude-code#47375

## Environment

- **LiteLLM version**: v1.88.0-rc.1 (Docker: `ghcr.io/berriai/litellm:v1.88.0-rc.1`)
- **Claude Code CLI version**: 2.1.159
- **Provider**: AWS Bedrock (Converse API)
- **Model**: `bedrock/converse/arn:aws:bedrock:ap-south-1:<account-id>:inference-profile/global.anthropic.claude-sonnet-4-6`
- **Region**: ap-south-1 (Mumbai) using CRIS global inference profile
- **OS**: Windows (Claude Code CLI) → LiteLLM on EC2 (Amazon Linux, ap-south-1 private subnet)
- **Thinking**: Enabled (extended thinking via Claude Code default)

## LiteLLM Config

```yaml
model_list:
  - model_name: claude-sonnet-4-6
    litellm_params:
      model: bedrock/converse/arn:aws:bedrock:ap-south-1:<account-id>:inference-profile/global.anthropic.claude-sonnet-4-6
      aws_region_name: ap-south-1

litellm_settings:
  drop_params: true
  modify_params: true
```

## Claude Code Settings

```json
{
  "env": {
    "ANTHROPIC_BASE_URL": "http://<litellm-proxy-ip>:4000",
    "ANTHROPIC_API_KEY": "sk-<key>"
  },
  "model": "claude-sonnet-4-6"
}
```


The error does NOT happen on every 2nd turn — it's intermittent (~30-50% of the time).

## Expected Behavior

The second turn should stream successfully. The thinking blocks from the previous turn should be properly handled in the Converse→Messages response translation.

## Suspected Root Cause

When LiteLLM translates the Bedrock Converse streaming response back to Anthropic `/v1/messages` format on multi-turn conversations:
- The conversation history contains thinking blocks from the previous assistant response
- The streaming response on the 2nd turn occasionally produces a content block event where a thinking block is positioned where Claude Code's streaming parser expects a text block
- This may be due to the ordering of content_block_start/content_block_delta SSE events during Converse→Messages translation


## Additional Context

- The error never shows in LiteLLM container logs (it's a client-side parsing failure)
- The LiteLLM container returns HTTP 200 with streaming — the issue is in the SSE event structure
- The error does NOT occur when using Anthropic's API directly (only through LiteLLM + Bedrock Converse)
- Previously tested with v1.84.4, v1.86.2 — same issue persists across versions


### Steps to Reproduce

1. 
## Steps to Reproduce

1. Start Claude Code CLI with the above config pointing to LiteLLM proxy
2. Send any prompt (first turn) — **works fine**
3. Send a follow-up prompt (second turn) — **intermittently fails** with "Content block is not a text block"
2. 
3. 


### Relevant log output

```shell
## Debug Log (from Claude Code CLI with `CLAUDE_CODE_DEBUG=1`)


16:13:39.138 [DEBUG] [API REQUEST] /v1/messages source=repl_main_thread    ← 1st turn
16:13:41.632 [DEBUG] Stream started - received first chunk                  ← ✅ works
16:13:46.965 [DEBUG] [API REQUEST] /v1/messages source=generate_session_title
16:13:47.065 [DEBUG] [API REQUEST] /v1/messages source=repl_main_thread    ← 2nd turn
16:13:48.150 [DEBUG] Stream started - first byte after 1114ms               (title gen)
16:13:49.730 [DEBUG] Stream started - first byte after 2688ms               (2nd turn)
16:13:50.728 [ERROR] Error in API request: Content block is not a text block
16:13:50.734 [ERROR] Error: Error: Content block is not a text block
    at TK_ (B:/~BUN/root/src/entrypoints/cli.js:10166:11320)
    at processTicksAndRejections (native:7:39)
```

### What part of LiteLLM is this about?

Proxy

### What LiteLLM version are you on ?

v1.84.4, v1.86.2, v1.88.0-rc.1

### Twitter / LinkedIn details

https://www.linkedin.com/in/dhinesh-raja-m-0135ab153/

--- a/litellm/llms/anthropic/experimental_pass_through/adapters/streaming_iterator.py
+++ b/litellm/llms/anthropic/experimental_pass_through/adapters/streaming_iterator.py
@@ -458,11 +458,14 @@ class AnthropicStreamWrapper(AdapterCompletionStreamWrapper):
 
                     # 3. If the trigger chunk carries delta content, queue it
                     # so the first delta of the new block is not silently dropped.
-                    if self._trigger_delta_has_content(processed_chunk):
+                    if self._delta_has_content(processed_chunk):
                         self.chunk_queue.append(processed_chunk)
 
                     self.sent_content_block_finish = False
                     return self.chunk_queue.popleft()
+
+                if processed_chunk["type"] == "content_block_delta" and not self._delta_has_content(processed_chunk):
+                    continue
 
                 if processed_chunk["type"] == "message_delta" and self.sent_content_block_finish is False:
                     # Queue both the content_block_stop and the message_delta
@@ -670,12 +673,17 @@ class AnthropicStreamWrapper(AdapterCompletionStreamWrapper):
 
                         # 3. If the trigger chunk carries delta content, queue it
                         # so the first delta of the new block is not silently dropped.
-                        if self._trigger_delta_has_content(processed_chunk):
+                        if self._delta_has_content(processed_chunk):
                             self.chunk_queue.append(processed_chunk)
 
                         # Reset state for new block
                         self.sent_content_block_finish = False
                         return self.chunk_queue.popleft()
+
+                    if processed_chunk["type"] == "content_block_delta" and not self._delta_has_content(
+                        processed_chunk
+                    ):
+                        continue
 
                     if processed_chunk["type"] == "message_delta" and self.sent_content_block_finish is False:
                         # Queue both the content_block_stop and the holding chunk
@@ -808,15 +816,22 @@ class AnthropicStreamWrapper(AdapterCompletionStreamWrapper):
         self.current_content_block_index += 1
 
     @staticmethod
-    def _trigger_delta_has_content(processed_chunk: Dict[str, Any]) -> bool:
-        """Return True if a translated trigger chunk carries a non-empty
-        ``content_block_delta`` payload that must be re-emitted after a
-        block transition.
+    def _delta_has_content(processed_chunk: Dict[str, Any]) -> bool:
+        """Return True if a translated chunk carries a non-empty
+        ``content_block_delta`` payload.
 
-        When an upstream chunk both *triggers* a new content block (its type
-        differs from the active block) and *carries* delta content, that
-        content belongs to the new block. The synthesized
-        ``content_block_start`` only ever carries an empty body — see
+        Gates every ``content_block_delta`` emission. An empty delta carries
+        no information, and the translate fallback types empty deltas as
+        ``text_delta`` regardless of the active block's type — emitting one
+        into an open ``thinking`` block (e.g. Bedrock Converse sends an empty
+        reasoning delta mid-block) crashes strict Anthropic SDK clients with
+        "Content block is not a text block".
+
+        Also gates re-emission after a block transition: when an upstream
+        chunk both *triggers* a new content block (its type differs from the
+        active block) and *carries* delta content, that content belongs to
+        the new block. The synthesized ``content_block_start`` only ever
+        carries an empty body — see
         ``_translate_streaming_openai_chunk_to_anthropic_content_block``,
         which returns an empty ``TextBlock``/``ToolUseBlock``/thinking block —
         so the trigger chunk's delta must be re-queued or the first token of

--- a/litellm/llms/anthropic/experimental_pass_through/adapters/streaming_iterator.py
+++ b/litellm/llms/anthropic/experimental_pass_through/adapters/streaming_iterator.py
@@ -13,7 +13,10 @@ from typing import (
     List,
     Literal,
     Optional,
+    get_args,
 )
+
+from typing_extensions import assert_never
 
 from litellm._logging import verbose_logger
 from litellm._uuid import uuid
@@ -21,6 +24,7 @@ from litellm.types.llms.anthropic import (
     AppliedEdit,
     CompactionBlock,
     ContextManagementResponse,
+    StreamingContentBlockDeltaType,
     UsageDelta,
     UsageIteration,
 )
@@ -28,6 +32,23 @@ from litellm.types.utils import AdapterCompletionStreamWrapper
 
 if TYPE_CHECKING:
     from litellm.types.utils import ModelResponseStream
+
+
+_STREAMING_DELTA_TYPES = frozenset(get_args(StreamingContentBlockDeltaType))
+
+
+def _delta_payload_field(delta_type: StreamingContentBlockDeltaType) -> str:
+    match delta_type:
+        case "text_delta":
+            return "text"
+        case "input_json_delta":
+            return "partial_json"
+        case "thinking_delta":
+            return "thinking"
+        case "signature_delta":
+            return "signature"
+        case _:
+            assert_never(delta_type)
 
 
 class _CombinedChunkSplitter:
@@ -837,6 +858,12 @@ class AnthropicStreamWrapper(AdapterCompletionStreamWrapper):
         so the trigger chunk's delta must be re-queued or the first token of
         the new block (the first non-empty text/thinking delta, or bundled
         tool arguments) is silently dropped.
+
+        Delta types outside ``StreamingContentBlockDeltaType`` — the closed
+        set the translate layer can produce — are treated as empty. The
+        per-type payload lookup is exhaustively matched against that set in
+        ``_delta_payload_field``, so extending the translate layer with a new
+        delta type fails type-checking here until it is handled.
         """
         if processed_chunk.get("type") != "content_block_delta":
             return False
@@ -844,15 +871,9 @@ class AnthropicStreamWrapper(AdapterCompletionStreamWrapper):
         if not isinstance(delta, dict):
             return False
         delta_type = delta.get("type")
-        if delta_type == "text_delta":
-            return bool(delta.get("text"))
-        if delta_type == "input_json_delta":
-            return bool(delta.get("partial_json"))
-        if delta_type == "thinking_delta":
-            return bool(delta.get("thinking"))
-        if delta_type == "signature_delta":
-            return bool(delta.get("signature"))
-        return False
+        if delta_type not in _STREAMING_DELTA_TYPES:
+            return False
+        return bool(delta.get(_delta_payload_field(delta_type)))
 
     def _should_start_new_content_block(self, chunk: "ModelResponseStream") -> bool:
         """

--- a/litellm/llms/anthropic/experimental_pass_through/adapters/transformation.py
+++ b/litellm/llms/anthropic/experimental_pass_through/adapters/transformation.py
@@ -104,6 +104,7 @@ from litellm.types.llms.anthropic import (
     ContextManagementResponse,
     MessageBlockDelta,
     MessageDelta,
+    StreamingContentBlockDeltaType,
     UsageDelta,
     UsageIteration,
 )
@@ -1423,7 +1424,7 @@ class LiteLLMAnthropicMessagesAdapter:
     def _translate_streaming_openai_chunk_to_anthropic(
         self, choices: List[Union[OpenAIStreamingChoice, StreamingChoices]]
     ) -> Tuple[
-        Literal["text_delta", "input_json_delta", "thinking_delta", "signature_delta"],
+        StreamingContentBlockDeltaType,
         Union[
             ContentTextBlockDelta,
             ContentJsonBlockDelta,

--- a/litellm/types/llms/anthropic.py
+++ b/litellm/types/llms/anthropic.py
@@ -439,6 +439,9 @@ class ContentThinkingSignatureBlockDelta(TypedDict):
     signature: str
 
 
+StreamingContentBlockDeltaType = Literal["text_delta", "input_json_delta", "thinking_delta", "signature_delta"]
+
+
 class ContentBlockDelta(TypedDict):
     type: Literal["content_block_delta"]
     index: int

--- a/tests/test_litellm/llms/anthropic/experimental_pass_through/adapters/test_streaming_iterator_first_delta.py
+++ b/tests/test_litellm/llms/anthropic/experimental_pass_through/adapters/test_streaming_iterator_first_delta.py
@@ -12,6 +12,13 @@ silently dropped — e.g. text resuming after a tool call started from the secon
 token ("The weather is nice." was lost, "Hi" rendered as ""). Bundled
 ``input_json_delta`` tool arguments were already preserved and must stay
 preserved, and empty trigger deltas must not produce spurious events.
+
+Also covers the inverse regression: a chunk whose translated delta carries no
+payload must not be emitted at all. The translate fallback types empty deltas
+as ``text_delta`` regardless of the active block, so an empty reasoning delta
+mid-thinking-block (Bedrock Converse sends these) used to emit ``text_delta``
+into an open ``thinking`` block, crashing Anthropic SDK clients (Claude Code)
+with "Content block is not a text block".
 """
 
 import os
@@ -49,6 +56,13 @@ def _make_chunk(delta: Delta, finish_reason: Optional[str] = None) -> MagicMock:
     chunk.usage = None
     chunk._hidden_params = {}
     return chunk
+
+
+def _thinking_chunk(thinking: str, signature: str = "") -> MagicMock:
+    block = {"type": "thinking", "thinking": thinking}
+    if signature:
+        block["signature"] = signature
+    return _make_chunk(Delta(content=None, thinking_blocks=[block]))
 
 
 def _tool_chunk(
@@ -107,6 +121,47 @@ def _input_json_deltas(events: List[dict]) -> List[str]:
         if e.get("type") == "content_block_delta"
         and e["delta"].get("type") == "input_json_delta"
     ]
+
+
+def _thinking_deltas(events: List[dict]) -> List[str]:
+    return [
+        e["delta"]["thinking"]
+        for e in events
+        if e.get("type") == "content_block_delta"
+        and e["delta"].get("type") == "thinking_delta"
+    ]
+
+
+def _signature_deltas(events: List[dict]) -> List[str]:
+    return [
+        e["delta"]["signature"]
+        for e in events
+        if e.get("type") == "content_block_delta"
+        and e["delta"].get("type") == "signature_delta"
+    ]
+
+
+_DELTA_TYPES_PER_BLOCK_TYPE = {
+    "text": {"text_delta"},
+    "thinking": {"thinking_delta", "signature_delta"},
+    "tool_use": {"input_json_delta"},
+}
+
+
+def _assert_deltas_match_their_block_type(events: List[dict]) -> None:
+    """Enforce the invariant the Anthropic SDK enforces client-side: every
+    ``content_block_delta`` must be of a type valid for the block opened by
+    the most recent ``content_block_start`` at the same index.
+    """
+    block_types = {}
+    for event in events:
+        if event.get("type") == "content_block_start":
+            block_types[event["index"]] = event["content_block"]["type"]
+        if event.get("type") == "content_block_delta":
+            block_type = block_types[event["index"]]
+            assert event["delta"]["type"] in _DELTA_TYPES_PER_BLOCK_TYPE[block_type], (
+                f"{event['delta']['type']} emitted into a {block_type} block: {event}"
+            )
 
 
 def test_held_stop_reason_usage_merge_preserves_openai_cache_token_details():
@@ -332,11 +387,70 @@ def test_bundled_tool_args_on_transition_still_preserved_sync():
         ({"type": "content_block_delta", "delta": None}, False),
     ],
 )
-def test_trigger_delta_has_content_branches(processed_chunk, expected):
-    """Directly exercise the re-emit predicate across all delta types and the
+def test_delta_has_content_branches(processed_chunk, expected):
+    """Directly exercise the emission predicate across all delta types and the
     empty/malformed guards, so the helper's behavior is pinned independently of
     upstream chunk-translation details.
     """
-    assert (
-        AnthropicStreamWrapper._trigger_delta_has_content(processed_chunk) is expected
+    assert AnthropicStreamWrapper._delta_has_content(processed_chunk) is expected
+
+
+def _empty_reasoning_delta_mid_thinking_chunks() -> List[MagicMock]:
+    return [
+        _thinking_chunk("Let me think"),
+        _thinking_chunk(""),
+        _thinking_chunk("", signature="sig123"),
+        _make_chunk(Delta(content="Hello")),
+        _make_chunk(Delta(content=None), finish_reason="stop"),
+    ]
+
+
+def _assert_empty_reasoning_delta_suppressed(events: List[dict]) -> None:
+    _assert_deltas_match_their_block_type(events)
+    assert _thinking_deltas(events) == ["Let me think"]
+    assert _signature_deltas(events) == ["sig123"]
+    assert _text_deltas(events) == ["Hello"]
+
+
+def test_empty_reasoning_delta_mid_thinking_block_is_suppressed_sync():
+    """Bedrock Converse repro: an empty reasoning delta arriving inside an open
+    thinking block used to be emitted as ``text_delta {"text": ""}`` at the
+    thinking block's index (no block transition), which crashes Claude Code's
+    Anthropic SDK with "Content block is not a text block". It must be dropped,
+    while the surrounding thinking/signature/text deltas all still flow.
+    """
+    wrapper = AnthropicStreamWrapper(
+        completion_stream=iter(_empty_reasoning_delta_mid_thinking_chunks()),
+        model="claude-x",
     )
+    _assert_empty_reasoning_delta_suppressed(_drain_sync(wrapper))
+
+
+@pytest.mark.asyncio
+async def test_empty_reasoning_delta_mid_thinking_block_is_suppressed_async():
+    """Async twin of the Bedrock Converse repro — the proxy serves the async
+    iterator, so the skip must exist on that path too.
+    """
+    wrapper = AnthropicStreamWrapper(
+        completion_stream=_AsyncStream(_empty_reasoning_delta_mid_thinking_chunks()),
+        model="claude-x",
+    )
+    _assert_empty_reasoning_delta_suppressed(await _drain_async(wrapper))
+
+
+def test_empty_content_chunk_mid_text_block_is_suppressed_sync():
+    """An empty-content chunk arriving mid-text-block (no transition) used to
+    emit a pointless ``text_delta {"text": ""}``; it must be dropped without
+    affecting the surrounding text deltas.
+    """
+    chunks = [
+        _make_chunk(Delta(content="Hi")),
+        _make_chunk(Delta(content="")),
+        _make_chunk(Delta(content=" there")),
+        _make_chunk(Delta(content=None), finish_reason="stop"),
+    ]
+    wrapper = AnthropicStreamWrapper(completion_stream=iter(chunks), model="claude-x")
+    events = _drain_sync(wrapper)
+
+    assert _text_deltas(events) == ["Hi", " there"]
+    _assert_deltas_match_their_block_type(events)

